Check target gates are parameterized in optimize1qgatesdecomposition

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit updates the Optimize1qGatesDecomposition transpiler pass to check the defined operation in the target to ensure it's a valid gate to use for the decomposition. Previously the pass would just use the gate name for each qubit and treat that as a basis gate list. However this ignored the extra details that the target provides as a target gate can match the name but not be valid for use in synthesis. This commit fixes that oversight by checking the gates directly and ensuring they're valid for use in the decomposition.

### Details and comments

Fixes #14743

mtreinish committed:
Check target gates are parameterized in optimize1qgatesdecomposition

This commit updates the Optimize1qGatesDecomposition transpiler pass to
check the defined operation in the target to ensure it's a valid gate to
use for the decomposition. Previously the pass would just use the gate
name for each qubit and treat that as a basis gate list. However this
ignored the extra details that the target provides as a target gate can
match the name but not be valid for use in synthesis. This commit fixes
that oversight by checking the gates directly and ensuring they're valid
for use in the decomposition.

Fixes #14743

mtreinish commented:
I didn't tag this for automatic backport because it won't apply cleanly to `stable/1.4`. The code still exists in 1.4, but it's in a different file that has been moved around on main since. Once this merges I'll manually backport it for inclusion in a 1.4.x bugfix release.

jakelishman commented:
Do we still want "stable backport potential" to trigger the backport to 2.2?

mtreinish commented:
Oh good point, I was only thinking about 1.4 but this will apply cleanly to 2.2 and fix it there.

jakelishman commented on crates/transpiler/src/passes/optimize_1q_gates_decomposition.rs:
```diff
@@ -76,7 +76,40 @@ pub fn run_optimize_1q_gates_decomposition(
+                                    .iter()
+                                    .all(|x| matches!(x, Param::ParameterExpression(_)))
+                                {
+                                    Some(gate.operation.name())
```
I think this bit of the logic is actually additional on top of the bugfix, right?

I generally agree it's a good idea to do this to communicate internally within the transpiler to extend the standard-gate handling to a larger range of "equivalent to standard gate" inputs but:
1. does allowing this case risk running counter to our (under-documented) assertions that the standard-gate names are special/reserved and should be used when representing standard gates?
2. is it worth doing this, when (I think) subsequent `instruction_supported` checks would fail on the outputs currently?
3. should we add a test for this part of the logic?

mtreinish replied:
That's a fair concern. I just figured this logic is what made the most sense, but I guess in a strict sense it could be viewed as behavior change for a backport (I think it would be fine in a minor version) since this could result in different gates being emitted in edge cases such as `Target.add_instruction(XGate(), name="Jake's Special X")` this would result in x gate being used when it wasn't before. I can change the logic here to not translate the `gate_name` to `gate.operation.name()` to make it fine for backport and do the translation as a follow-up only on main.

I will say the potential change here is actually super small, it's only for a subset of our 1q gates and when they're added to the target with a custom name.

jakelishman replied:
I am generally on board with the new logic taken entirely in isolation.  My worries are more about that I'm not sure if it fits in with other messaging/assertions about "standard gates are reserved" (I don't really have a clear view of what and when we require things in this model), and that I don't think this is the only place where it'd have an impact (for example, consider the choice of Euler decomposer within `UnitarySynthesis`), so it seems weird to have it in just one place, especially for logic added in a backport.

I think this trouble of the name/gate mismatch is rubbing against the general problems in the `Target` model we've talked about, where it's possible to get from a `Target` to a set of supported instructions, but it's hard/expensive to thoroughly check a circuit instruction against the `Target` - in this case, we're not checking a circuit instruction, but kind of checking for implicit instances of the standard gates and hitting related problems around the name lookups not necessarily matching.

mtreinish replied:
Ok I removed the behavior change here in: https://github.com/Qiskit/qiskit/pull/15131/commits/d85c431f418dddeafff4e1f28229a9cd4abc7c52 and just made it straight filter which will just fix the bug without any other changes.

jakelishman commented on releasenotes/notes/fix-1q-decomp-non-param-gates-99624b82227e5608.yaml:
```diff
@@ -0,0 +1,9 @@
+fixes:
+  - |
+    Fixed an issue in the :class:`.Optimize1qGatesDecomposition` when the
+    pass was initialized with a target that contains 1q gates with fixed angle
```
```suggestion
    pass was initialized with a :class:`.Target` that contains 1q gates with fixed angle
```

jakelishman commented on releasenotes/notes/fix-1q-decomp-non-param-gates-99624b82227e5608.yaml:
```diff
@@ -0,0 +1,9 @@
+  - |
+    Fixed an issue in the :class:`.Optimize1qGatesDecomposition` when the
+    pass was initialized with a target that contains 1q gates with fixed angle
+    parameters. Previously, the pass would potential output gates outside
```
```suggestion
    parameters. Previously, the pass would potentially output gates outside
```

mtreinish committed:
Fix release note typos

Co-authored-by: Jake Lishman <jake@binhbar.com>

mtreinish committed:
Simplify logic in pass

Since this PR is intended to be backported to 2.2.x and 1.4.x we should
minimize the diff. The previous version of this PR changed the behavior
of the pass to normalize on using standard gates even if there were
custom names for gates in the target. Longer term I think this behavior
is better for the pass, but there is discussions around exactly the
semantics of the Target's data model around names vs operations here and
also that this might change the output pretty noticeably in some cases.
For a bugfix intended for backport it's better to be conservative here,
so this commit reverts the logic to just use the name in the target and
only filter operations if the target doesn't actually support arbitrary
angles on them. We can revisit the improvements in a follow up PR.

jakelishman reviewed (approved):
It still might be worth revisiting the new logic that was in the original draft of this PR, but I agree this is safer for backporting.
